### ISSUE TITLE
Quasar: gtz/gez/ltz/lez return 1 for NaN — the comparison-to-zero predicate has no inf term, which Wormhole and Blackhole both carry

### ISSUE THREAD

**Issue by singhharsh1708** [tenstorrent/tt-metal#55273]
### Summary

On Quasar, `ttnn.gtz(nan)` and `ttnn.gez(nan)` return `1.0`, and `ttnn.ltz(-nan)` / `ttnn.lez(-nan)` return `1.0`. All four should be `0.0`, and are on Wormhole B0 and Blackhole. The comparison-to-zero predicate classifies a lane from the sign bit and "magnitude != 0" only, with no term to reject NaN.

This is the same shape as #54857: a correctness fix present on Wormhole and Blackhole that the Quasar copy never received.

### The Quasar predicate

`tt_metal/hw/ckernels/quasar/metal/llk_api/llk_sfpu/ckernel_sfpu_comp.h:118-132`:

```cpp
const sfpi::vSMag mag = sfpi::setsgn(sfpi::as<sfpi::vSMag>(v), 0);
...
} else if constexpr (COMP_MODE == SfpuType::less_than_zero) {
    return (v < 0) && (mag != 0);   // sign set and nonzero -> excludes -0.0
} else if constexpr (COMP_MODE == SfpuType::greater_than_zero) {
    return (v >= 0) && (mag != 0);  // sign clear and nonzero
} else if constexpr (COMP_MODE == SfpuType::greater_than_equal_zero) {
    return (v < 0) && (mag != 0);   // loop defaults 1 and writes 0 here -> gtez
} else {                            // less_than_equal_zero
    return (v >= 0) && (mag != 0);  // loop defaults 1 and writes 0 here -> ltez
}
```

For `+NaN` = `0x7FC00000`: bit 31 is clear so `v >= 0` holds, and `mag = 0x7FC00000 != 0`, so `greater_than_zero` returns true and the lane is written `1`. The `greater_than_equal_zero` write-0 predicate `(v < 0) && (mag != 0)` is false for the same value, so that lane keeps the loop default of `1`. `-NaN` = `0xFFC00000` breaks `ltz` and `lez` symmetrically.

### Wormhole and Blackhole reject NaN explicitly

`tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_comp.h:78-131` loads the inf bit pattern and adds a compare against it, with the intent stated in the comments:

```cpp
constexpr std::uint32_t INF = p_sfpu::LREG5;
constexpr std::uint32_t BFLOAT16_INF = 0x7f80;
...
// ltz: default 0; chain: (v < 0) AND (|v| != 0) AND (|v| <= inf) -> 1; NaN: |NaN| > inf -> rejected
TTI_SFPIADD(0, INF, ABS_V, sfpi::SFPIADD_MOD1_ARG_2SCOMP_LREG_DST | sfpi::SFPIADD_MOD1_CC_GTE0);
...
// gtz: default 0; chain: (v >= 0) AND (|v| != 0) AND (|v| <= inf) -> 1; NaN: |NaN| > inf -> rejected
...
// gez: default 1; chain1: (v<0) AND (|v|!=0) -> 0 (negatives excl. -0); chain2: |v|>inf -> 0 (NaN)
```

Blackhole has the equivalent at `ckernel_sfpu_comp.h:54,59,83-122`. Only Quasar omits it.

`+/-inf` (`0x7F800000` / `0xFF800000`) have magnitude exactly equal to inf and behave identically on all three architectures — only NaN diverges.

### Reach

`tt_metal/hw/inc/api/compute/eltwise_unary/comp.h` routes `gtz_tile` / `gez_tile` / `ltz_tile` / `lez_tile` to `sfpu::calculate_zero_comp` under `#ifdef ARCH_QUASAR`, where the `#ifndef` branch uses the Wormhole/Blackhole `calculate_comp`. `unary_op_utils.cpp:871-897` emits those calls for `UnaryOpType::GTZ/GEZ/LTZ/LEZ` on non-int32 dtypes, and `unary_nanobind.cpp:1484,1490,1539,1551` bind `ttnn.gez`, `ttnn.gtz`, `ttnn.lez`, `ttnn.ltz`.

The same four are also the postprocess step for the FPU path of `ttnn.gt` / `ge` / `lt` / `le` (`binary_ng_utils.cpp:195-206`), which compute `a - b` and then compare to zero. A NaN operand propagates a NaN into that subtraction, so those comparisons return True where they should return False — a `where`/masking predicate then selects the wrong branch silently.

### Expected vs actual

```python
ttnn.gtz(ttnn.from_torch(torch.tensor([float('nan')], dtype=torch.bfloat16), layout=ttnn.TILE_LAYOUT, device=dev))
```

`torch.gt(nan, 0)` is False, `torch.ge(nan, 0)` is False. Wormhole and Blackhole return `0.0`. Quasar returns `1.0`. Exactly two of the four sign-sensitive comparisons are wrong for any given NaN sign.

### Fix

Add the missing inf term to the four sign-sensitive arms at `quasar/.../ckernel_sfpu_comp.h:125-131`, mirroring the Wormhole/Blackhole chain. With `const sfpi::vSMag inf = sfpi::as<sfpi::vSMag>(sfpi::vInt(0x7F800000));` alongside the existing `mag`:

- `less_than_zero` → `(v < 0) && (mag != 0) && (mag <= inf)`
- `greater_than_zero` → `(v >= 0) && (mag != 0) && (mag <= inf)`
- `greater_than_equal_zero` (write-0) → `((v < 0) && (mag != 0)) || (mag > inf)`
- `less_than_equal_zero` (write-0) → `((v >= 0) && (mag != 0)) || (mag > inf)`

`equal_zero` / `not_equal_zero` are unchanged — they are already correct for NaN. This is a predicate edit inside the existing `v_if`: no new kernel, no extra dispatch.

Quasar is pre-silicon, so as with #54857 I cannot run this; the analysis is a static comparison of the three architectures' kernels against `2aa657a`, and someone with Quasar access should confirm the boundary case.

I would like to work on this through the bounty program, and am happy to prepare the PR.

**Comment by badnikhil** [comment]
Confirmed against current main: `_zero_comp_pred_` in `quasar/.../ckernel_sfpu_comp.h` has no inf/NaN term, so `+NaN` satisfies the gtz predicate (sign clear, magnitude nonzero) and `-NaN` the ltz one; the write-0 chains for gez/lez miss NaN for the same reason.

I have a fix in #55677 that adds the missing term to the four sign-sensitive float predicates (NaN = any magnitude above the inf bit pattern, integer compare, mirroring the WH/BH `|v| <= inf` chains) and leaves eqz/nez and all integer formats byte-identical. `test_sfpu_compute.cpp` now pins ±NaN/±inf/±0 lanes in the comparison-to-zero stimuli. Without Quasar silicon (and ttnn ops aborting on `libttsim_qsr.so` before any kernel runs), verification is a host-side SFPI compile of all 36 instantiations with the tt-llk Quasar harness flags, an objdump diff showing only those four instantiations change, and the same inputs on ttsim Blackhole returning 0 for both NaN signs. Happy to have this considered for the bounty program if the maintainers think it qualifies.

**Comment by singhharsh1708** [comment]
Thanks for confirming it. Your point that the write-0 chains for `gez` and `lez` miss NaN for the same reason is the piece I had not spelled out, so it is one missing inf term rather than two separate faults.

For context, I reported this one and I am working on a fix for it from the ttsim side, so I would rather we did not duplicate the work.

Happy to compare notes on your fix rather than have two in flight for the same report.

**Comment by badnikhil** [comment]
Fair — you reported it, so I am not contesting the bounty; if your fix is in progress it should stay with your report.

#55677 stays open as reference only. The bits worth reusing: `_zero_comp_pred_` needs an `IS_FLOAT` parameter because sfpi has no `>` or `<=` on integers, so the NaN term has to be written as `mag >= 0x7F800001`; only the four sign-sensitive Float32 instantiations change, and eqz/nez plus every integer format stay byte-identical (I checked with an objdump diff).

Since Quasar cannot execute anything on `libttsim_qsr.so` yet, the only verification I could do was a host-side SFPI compile of all 36 instantiations with the tt-llk Quasar flags plus the Blackhole reference behaviour — worth knowing if you plan to measure it directly.

Say the word and I will close it.
